### ISSUE TITLE
Fatal Errors when using 'Future Mode' in v2.0.0

### ISSUE THREAD

**Issue by simplechris** [elastic/elasticsearch-php#218]
I've been running into some problems with the async stuff the new version of this library, refered to in the docs as [Future Mode](http://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_future_mode.html). I had initially suspected it was something to do with PHP 5.4, but i'm experiencing the same issue in 5.5 and 5.6 environments.

The problem embodies itself as a Fatal Error, caused by an uncaught `Elasticsearch\Common\Exceptions\Serializer\JsonErrorException` exception, and by digging into it, its because the Serializer instance is passed malformed JSON. 

Here's some code that causes the error, so you can hopefully recreate it:

- Run this file [`index.php`](https://gist.github.com/simplechris/fc50d21c5788833c5119#file-index-php)
- Here is the error output I get when I run the above file: [`error-output.txt`](https://gist.github.com/simplechris/fc50d21c5788833c5119#file-output-txt)
- Here's my phpinfo, which might be helpful with debugging: [`phpinfo() output`](https://gist.github.com/simplechris/edfb0274beee58c93522)

I know that ES isnt returning the malformed json, so something within this library or within guzzle is messing up the stream that the data is written into.

I believe this bug could be caused by the various handles been managed by a `curl_multi_exec` call all writing into (the same?) [temporary stream](https://github.com/elastic/elasticsearch-php/blob/6cbe2fe048d70d80feff4674cab739d72fc61f0c/src/Elasticsearch/ClientBuilder.php#L369), that seems to get `rewind`ed (rewound?) within [`performRequest()`](https://github.com/elastic/elasticsearch-php/blob/6cbe2fe048d70d80feff4674cab739d72fc61f0c/src/Elasticsearch/Connections/Connection.php#L161). could this result in a stream of malformed json (with new calls overwriting the existing data the start of the stream?) I could also be going down the complete wrong path here...

I'm finding it fairly hard to debug, and I'm not particularly familiar with the Guzzle/Ring internals.

Any ideas?

**Comment by polyfractal** [comment]
Well, the TL;DR: I "fixed" it by removing all the `save_to` stream functionality.  I just pushed a commit and it no longer errors on my side...does it work for you too?

Your intuition was correct, the stream stuff can corrupt the JSON if everything isn't rewound correctly, since there will be extra garbage in the stream.  I think the problem was this:

1. We give RingPHP a temp stream to use
2. Before each handler invocation, we rewind the stream
3. RingPHP queues up a bunch of requests in for parallel execution
4. Ring executes first query, appends to stream from [0..q1]
5. Ring executes second query, **appends to stream from [q1..q2]**.  The rewind happens *before* any of the queries are executed, not after each execution.  So all responses are stacked in the stream 
6. When responses come back, we try to rewind to zero each time and pull content-length out of the stream...which is the first response repeatedly.  As soon as there is a different length we'll throw an invalid JSON error.

You can see it happening by using a function_score with random_score ... the first score is always identical, meaning it's just pulling the response each time:

```php
'body'  => [
    'query' => [
        'function_score' => [
            'functions' => [
                ['random_score'=>['seed'=> $i]]
            ]
        ]
    ]
]
```
```
>>>1690
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}}]}}

>>>1622
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_source":{"timestamp": "201

>>>1624
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_source":{"timestamp": "2014-

>>>1623
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_source":{"timestamp": "2014

>>>1596
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_

>>>1686
{"took":1,"timed_out":false,"_shards":{"total":5,"successful":5,"failed":0},"hits":{"total":402,"max_score":0.99971485,"hits":[{"_index":"test","_type":"test","_id":"AUy5d3OYkklL-_uwzYdB","_score":0.99971485,"_source":{"timestamp": "2014-12-5T00:00:00.000Z","price": 5}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeA","_score":0.9981775,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d2S1kklL-_uwzYa-","_score":0.9949093,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d10ekklL-_uwzYaE","_score":0.99422723,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcS","_score":0.99060214,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d2fvkklL-_uwzYbY","_score":0.99022764,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}},{"_index":"test","_type":"test","_id":"AUy5d31pkklL-_uwzYeQ","_score":0.98660195,"_source":{"timestamp": "2014-12-8T00:00:00.000Z","price": 8}},{"_index":"test","_type":"test","_id":"AUy5d23tkklL-_uwzYcT","_score":0.98488766,"_source":{"timestamp": "2014-12-9T00:00:00.000Z","price": 9}},{"_index":"test","_type":"test","_id":"AUy5d3wAkklL-_uwzYeB","_score":0.98365206,"_source":{"timestamp": "2014-12-2T00:00:00.000Z","price": 2}},{"_index":"test","_type":"test","_id":"AUy5d3akkklL-_uwzYdW","_score":0.9835514,"_source":{"timestamp": "2014-12-1T00:00:00.000Z","price": 1, "field": "abc", "field2": 123}
```

I didn't notice it because it works no problem in "non-async" mode, and in async mode it works for single requests.  And it *appears* to work for requests with identical length, although it's just repeating the same response over and over (which I wasn't looking for ... need to write a test for that).

So basically I didn't reason through the stream stuff very well :)  It can be fixed by not rewinding the response body, but this will allow the stream to grow infinitely until it starts to buffer to disk.  

I ripped all the stream stuff out in the https://github.com/elastic/elasticsearch-php/commit/f32e03800bf7580071bab11d15255eefe12b0a7e, because it's not really needed.  I originally put it in to support PHP streams, but ended up deciding that was a performance trap and users should just install libcurl.  The curl handlers don't need a `save_to`, so none of this stuff is necessary (and probably slower anyhow, since it needs several more allocations to get the data to the end-user).

Lemme know if this fixes it on your end :)

**Comment by Nickology** [comment]
Fixed it on my end for sure. Please release :smile:

**Comment by polyfractal** [comment]
Oops, forgot about this.  Will tag and release when I get to the airport tonight.  Thanks for the bump!

**Comment by simplechris** [comment]
:+1:

**Comment by Nickology** [comment]
Thanks!

**Comment by polyfractal** [comment]
@Nickology @simplechris Tagged as `v2.0.0-beta3`.  Sorry for the delay!
